D_CUT

ctzsnooze opened the pull request:
This code cuts D by a specified percentage durning normal flight, and lets D smoothly rise back to normal with rapid gyro moves like flips and rolls.  D will also increase little bit during prop wash events.
Users can now tune D in the 'normal' 35-40 range like in 3.5, but have the benefit of less D unless it is needed.

For example, if D is 40, a d_cut_percentage of 65 will cut D to 14 in normal flight, but the quad will still get full 40 of D to control bounce-back after flips and about 25 of D during strong prop wash.
The d_cut_percentage can be adjusted via the OSD, from the D filtering page.

Adding d_cut results in cooler motors, lower amounts of noise in motor traces in normal flight, and faster reactions to quick stick inputs.

Too much d_cut_percentage may result in not enough D generally, bringing out P oscillation if P is tuned quite high, but this has not been an issue to date.

Input to the boost signal is gyro differential (from gyro delta in the pid loop) with frequencies above 40hz (above propwash) attenuated with a configurable cutoff biquad filter (d_cut_range_hz).  Lower values of d_cut_range_hz, but probably not less than 30hz, can be used, if the quad is very noisy or gets low frequency D resonant oscillation.  Up to 50 or 60hz may be better on clean quads where prop wash control is the main priority.  Too high results in D being boosted from noise in normal flight.

The boost signal is 'integrated, smoothed and delayed' with a 7hz PT1 'd_cut_lowpass_hz' filter.  The default of 7Hz gives about the right amount of smoothing and delay.  Higher numbers cause the boost to come on faster, with less delay, but this is not usually desirable, since then the D boost starts to interfered with stick responsiveness, and can make the boost wobble around more than one might like.  This parameter could be removed from the CLI if users find the default is fine.

The gain control amplifies the input to the boosting effect.  Logging with set debug_mode = D_CUT allows recording of realtime D values on roll and pitch into debug 2 and 3.  If the user has quicker thumbs or a fast reacting quad, the gain control can be reduced.  Ideally the gain hits max D and sort of flat tops the boost around the time D would naturally peak.  Gain of 13 seems pretty good on most test machines. 

Typically this code doesn't do much except in the context of very rapid gyro moves, such as flips.  

For freestylers it allows quite high D, snappy entry into moves, and soft exits.  For racers I would not push D so high as a freestyler might.  That's because with repetitive fast inputs d will remain boosted and may dampen responsiveness, or smooth things out, a bit more than you might like, e.g. in slaloms or rapid zig-zags.  That's about the only downside.  Some people might like this, I don't know.  But the solution is to not go crazy on how high to push D, e.g. use D of say 30 with cut of 50% say.

This code works well with around 50% of D cut on TPA also, leaving as little as 6 of D on full throttle.  Unless P is over-tuned, this hasn't seemed to be a problem, and definitely cuts motor noise on full throttle significantly.  This has been tested by the usual suspects and a few ring-ins, so far all responses have been positive.

The downside is extra CPU from one more biquad and one more PT1, both at PID loop speed.

I am sure the coding can be improved and will be grateful for any suggestions to make it more efficient.

ctzsnooze commented:
A significant detail is that the default D values are back up to high values, since the Cut is enabled by default, which allows using relatively high D like we enjoyed in the past once again.  I'd be grateful for advice on whether these values are too high or not.  They should be fine on new installations if the cut is enabled for the target, but some targets won't support d_cut, and then they may well be too high.
Back in 3.2 we had D of 30 on Roll and 35 on Pitch, we now have them lower at 25 and 27, but many pilots would quite like higher D values if possible.  
I'm proposing 44 and 47, which perhaps may be a little on the high side.  
Maybe we could test like this, and then, based on more user feedback, see if this will be OK?

mikeller commented on src/main/flight/pid.h:
```diff
@@ -160,6 +161,10 @@ typedef struct pidProfile_s {
     uint8_t use_integrated_yaw;             // Selects whether the yaw pidsum should integrated
     uint8_t integrated_yaw_relax;           // Specifies how much integrated yaw should be reduced to offset the drag based yaw component
     uint8_t thrustLinearization;            // Compensation factor for pid linearization
+    uint8_t dterm_cut_percent;              // Amount to cut D by with no gyro activity, zero disables, 20 means cut 20%, 50 means cut 50%
```
These changes require the parameter group id to be incremented in `pid.c`, as this is an ARRAY type parameter group. (see https://github.com/betaflight/betaflight/blob/master/docs/development/ParameterGroups.md.)

ctzsnooze replied:
thanks! done.

mikeller commented:
This is overflowing RAM on LUXV2_RACE, so it needs to be either disabled for LUXV2_RACE, or only enabled for >= F4.

ctzsnooze commented on src/main/flight/pid.h:
```diff
@@ -39,6 +39,7 @@
 #define PTERM_SCALE 0.032029f
 #define ITERM_SCALE 0.244381f
 #define DTERM_SCALE 0.000529f
+#define D_CUT_GAIN_FACTOR 0.00005f
```
thanks, have done so, and put inside if defined conditionals (hopefully that is the right thing to do)

mikeller commented on src/main/target/LUX_RACE/target.h:
```diff
@@ -27,7 +27,7 @@
 
 // Removed to make the firmware fit into flash (in descending order of priority):
-//#undef USE_RTC_TIME
+#undef USE_RTC_TIME
```
Please keep pull requests focused on one change per pull request, and do not use them to change unrelated features.
Removing existing features from targets will be reported as 'bug' by users using these features, and having to dig through pull requests that don't clearly specify that they are doing this will just make this harder to support.

The easiest way to address this is to add an `#undef USE_D_CUT` to this target. Alternatively, features will have to be pruned from this target in a separate pull request.

ctzsnooze replied:
done

ctzsnooze commented on src/main/cms/cms_menu_imu.c:
```diff
@@ -586,6 +589,7 @@ static OSD_Entry cmsx_menuFilterPerProfileEntries[] =
     { "DTERM LPF2", OME_UINT16, NULL, &(OSD_UINT16_t){ &cmsx_dterm_lowpass2_hz,    0, 500, 1 }, 0 },
     { "DTERM NF",   OME_UINT16, NULL, &(OSD_UINT16_t){ &cmsx_dterm_notch_hz,       0, 500, 1 }, 0 },
     { "DTERM NFCO", OME_UINT16, NULL, &(OSD_UINT16_t){ &cmsx_dterm_notch_cutoff,   0, 500, 1 }, 0 },
+    { "DTERM CUT",  OME_UINT8,  NULL, &(OSD_UINT8_t) { &cmsx_dterm_cut_percent,    0, 100, 1 }, 0 },
```
I was wondering about that.  Have put all aspects of it into ifdef's, hope that's correct.

ctzsnooze commented:
Rebased and squashed.  Very slightly different defaults based on user feedback.  Thanks to everyone who helped develop and test this code, especially George and Ken.

BrenLaw commented:
@ctzsnooze You mentioned this works well with TPA, but what about Thrust Linearization? Should people using Thrust Linearization still run lower D terms than the newly raised defaults?

jcutcher01 commented:
I am running D=43 freestyle on Stingy and Alien frames. I look forward to testing this out.
